src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiConnection.cs

```diff
@@ -0,0 +1,239 @@
+            }
+            catch (Exception ex)
+            {
+                Logger.TryGet(LogEventLevel.Error, nameof(DxgiSwapchain))
+                    ?.Log(null, "Unable to establish Dxgi: {0}", ex);
+            }
+        }
+
+        private unsafe void RunLoop()
+        {
+            _stopwatch = System.Diagnostics.Stopwatch.StartNew();
+            try
+            {
+                GetBestOutputToVWaitOn();
+            }
+            catch (Exception ex)
+            {
+                Logger.TryGet(LogEventLevel.Error, nameof(DxgiSwapchain))
+                                    ?.Log(this, $"Failed to wait for vblank, Exception: {ex.Message}, HRESULT = {ex.HResult}");
+            }
+
+            while (true)
+            {
+                try
+                {
+                    lock (_syncLock)
+                    {
+                        if (_output is not null)
+                        {
+                            try
+                            {
+                                _output.WaitForVBlank();
+                            }
+                            catch (Exception ex)
+                            {
+                                Logger.TryGet(LogEventLevel.Error, nameof(DxgiSwapchain))
+                                    ?.Log(this, $"Failed to wait for vblank, Exception: {ex.Message}, HRESULT = {ex.HResult}");
+                                _output.Dispose();
+                                _output = null;
+                                GetBestOutputToVWaitOn();
+                            }
+                        }
+                        else
+                        {
+                            // well since that obviously didn't work, then let's use the lowest-common-denominator instead 
+                            // for reference, this has never happened on my machine,
+                            // but theoretically someone could have a weirder setup out there 
+                            DwmFlush();
+                        }
+                        Tick?.Invoke(_stopwatch.Elapsed);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Logger.TryGet(LogEventLevel.Error, nameof(DxgiSwapchain))
+                                    ?.Log(this, $"Failed to wait for vblank, Exception: {ex.Message}, HRESULT = {ex.HResult}");
+                }
+            }
+        }
+
+        // Note: Defining best as display with highest refresh rate on 
+        private void GetBestOutputToVWaitOn()
+        {
+            double highestRefreshRate = 0.0d;
+            HRESULT retval = default;
+            IDXGIFactory? fact = null;
+            void* factPointer = null;
+
+            // IDXGIFactory Guid: [Guid("7B7166EC-21C7-44AE-B21A-C9AE321AE369")]
+            Guid factoryGuid = MicroComRuntime.GetGuidFor(typeof(IDXGIFactory));
+            retval = CreateDXGIFactory(&factoryGuid, (void**)&factPointer);
```

Comment:
You could have declared it as 
```
[DllImport("dxgi", ExactSpelling = true, PreserveSig = false)]
internal static extern void CreateDXGIFactory(Guid* riid, void** ppFactory);
```
Exception for non-success HRESULT would be thrown automatically.

Reply:
```cs
[DllImport("dxgi", ExactSpelling = true, PreserveSig = false)]
internal static extern void* CreateDXGIFactory(Guid* riid);
```

works too, IIRC.

Reply:
These are good suggestions, I have taken them into the code.